src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -111,6 +114,167 @@ impl PlanTreeNode for LogicalMultiJoin {
+            output =
+                LogicalProject::with_mapping(output, self.mapping_from_ordering(join_ordering))
+                    .into();
+        }
+
+        // We will later push down all of the filters back to the individual joins via the
+        // `FilterJoinRule`.
+        output = LogicalFilter::create(output, self.on.clone());
+
+        output
+    }
+
+    // Our heuristic join reordering algorithm will try to perform a left-deep join.
+    // It will try to do the following:
+    //
+    // 1. First, split the join graph, with eq join conditions as graph edges, into their connected
+    //    components. Repeat the procedure in 2. with the largest connected components down to
+    //    the smallest.
+    // 2. For each connected component, add joins to the chain, prioritizing adding those
+    //    joins to the bottom of the chain if their join conditions have:
+    //       a. eq joins between primary keys on both sides
+    //       b. eq joins with primary keys on one side
+    //       c. more equijoin conditions
+    //    in that order. This forms our selectivity heuristic.
+    // 3. Thirdly, we will emit a left-deep cross-join of each of the left-deep joins of the
+    //    connected components. Depending on the type of plan, this may result in a planner failure
+    //    (e.g. for streaming). No cross-join will be emitted for a single connected component.
+    // 4. Finally, we will emit, above the left-deep join tree:
+    //        a. a filter with the non eq conditions
+    //        b. a projection which reorders the output column ordering to agree with the
+    //           original ordering of the joins.
+    //    The filter will then be pushed down by another filter pushdown pass.
+    pub(crate) fn heuristic_ordering(&self) -> Result<Vec<usize>> {
+        let mut labeller = ConnectedComponentLabeller::new(self.inputs.len());
+
+        let (eq_join_conditions, _) = self.on.clone().split_by_input_col_nums(
+            &self.input_col_nums(),
+            // only_eq=
+            true,
+        );
+
+        // Iterate over all join conditions, whose keys represent edges on the join graph
+        for k in eq_join_conditions.keys() {
+            labeller.add_edge(k.0, k.1);
+        }
+
+        let mut edge_sets: Vec<_> = labeller.into_edge_sets();
+
+        // Sort in decreasing order of len
+        edge_sets.sort_by_key(|a| std::cmp::Reverse(a.len()));
+
+        let mut join_ordering = vec![];
+
+        for component in edge_sets {
+            let mut eq_cond_edges: Vec<(usize, usize)> = component.into_iter().collect();
+
+            // TODO(jon-chuang): add sorting of eq_cond_edges based on selectivity here
+            eq_cond_edges.sort();
+
+            if eq_cond_edges.is_empty() {
+                // There is nothing to join in this connected component
+                break;
+            };
+
+            let edge = eq_cond_edges.remove(0);
```

Comment:
The loop looks like O(n^2), but I'd be okay with this as we won't have many join tables...

Reply:
yes, in practice, the number of join conditions also tends to be small.